staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go

```diff
@@ -27,6 +27,7 @@ import (
 // CustomResource is a JSON data structure.
 func ValidateCustomResource(customResource interface{}, validator *validate.SchemaValidator) error {
 	result := validator.Validate(customResource)
+	result.ApplyDefaults()
```

Comment:
Validation should not mutate

Reply:
+1.

BTW, I find we can set default values for `api/v1` resource in `pkg/api/v1/defaults.go`, is there any place where we can set default values for custom resource?

Reply:
For OpenAPI defaulting and validation are related. Once, an object is validated we have the defaulter.

Obviously, this does not match the order in kube where we default first during decoding and then validate the internal type. We can make this work by
- either accepting late defaulting in the validation steps (as done in this PR)
- or we do validation twice: first during defaulting, skipping defaulting on validation errors; second here in `ValidateCustomResource`.

Either way is kind of ugly. 

@liggitt any other idea or which of the two options (with sufficiently explicit comments around the uglyness) sounds like the lesser evil?

Reply:
> For OpenAPI defaulting and validation are related. Once, an object is validated we have the defaulter.

By the time we get here, it seems like this should be a pure validation pass (especially if multiple validation methods could be involved eventually)

Doing a defaulting pass first, earlier, seems like it would be easier to follow.